modules/PanelApp.py

```diff
@@ -0,0 +1,227 @@
+
+            # Remove empty or low quality records.
+            .filter(
+                ~(
+                    # There is neither a phenotype string nor an ontology identifier.
+                    ((col('phenotype').isNull()) & (col('diseaseFromSourceId').isNull())) |
+                    # The name of the phenotype string starts with a question mark, indicating low quality.
+                    col('phenotype').startswith('?')
+                )
+            )
+            .persist()
+        )
+
+        # Fetch and join literature references.
+        all_panel_ids = panelapp_df.select('Panel Id').toPandas()['Panel Id'].unique()
+        literature_references = self.fetch_literature_references(all_panel_ids)
+        panelapp_df = panelapp_df.join(literature_references, on=['Panel Id', 'Symbol'], how='left')
+
+        # Populate final evidence string structure.
+        panelapp_df = (
+            panelapp_df
+            # allelicRequirements requires a list, but we always only have one value from PanelApp.
+            .withColumn(
+                'allelicRequirements',
+                when(col('Mode of inheritance').isNotNull(), array(col('Mode of inheritance')))
+            )
+            .drop('Mode of inheritance')
+
+            .withColumnRenamed('List', 'confidence')
+            .withColumn('datasourceId', lit('genomics_england'))
+            .withColumn('datatypeId', lit('genetic_literature'))
+            .withColumnRenamed('phenotype', 'diseaseFromSource')
+            # diseaseFromSourceId populated above
+            # literature populated above
+            .withColumnRenamed('Panel Id', 'studyId')
+            .withColumnRenamed('Panel Name', 'studyOverview')
+            .withColumnRenamed('Symbol', 'targetFromSourceId')
+        )
+
+        # Save data.
+        with tempfile.TemporaryDirectory() as tmp_dir_name:
+            (
+                panelapp_df.coalesce(1).write.format('json').mode('overwrite')
+                .option('compression', 'org.apache.hadoop.io.compress.GzipCodec').save(tmp_dir_name)
+            )
+            json_chunks = [f for f in os.listdir(tmp_dir_name) if f.endswith('.json.gz')]
+            assert len(json_chunks) == 1, f'Expected one JSON file, but found {len(json_chunks)}.'
+            os.rename(os.path.join(tmp_dir_name, json_chunks[0]), output_file)
+
+    def fetch_literature_references(self, all_panel_ids):
+        """Queries the PanelApp API to extract all literature references for (panel ID, gene symbol) combinations."""
+        publications = []  # Contains tuples of (panel ID, gene symbol, PubMed ID).
+        for panel_id in all_panel_ids:
+            logging.debug(f'Fetching literature references for panel {panel_id!r}.')
+            url = f'https://panelapp.genomicsengland.co.uk/api/v1/panels/{panel_id}'
+            for gene in requests.get(url).json()['genes']:
+                for publication_string in gene['publications']:
```

Comment:
Something strange is happening here. Many of the genes have pubmed id `2`. Which is most likely an artefact. 

Let's say panel id: `234`, this function returns a dataframe in which the following rows exist:

```
+--------+------+-------------+
|Panel ID|Symbol|   literature|
+--------+------+-------------+
|     234| ACTG2|          [2]|
|     234|TBC1D1|          [2]|
|     234| GRIP1|[2, 24700879]|
|     234| SALL4|          [2]|
|     234| TRAP1|          [2]|
|     234| FREM1|          [2]|
+--------+------+-------------+
```
This is not good. If I take a look at the API response I see:
```
ACTG2   - ['PMID: 25998219']
FREM1   - ['PMID: 24700879']
GRIP1   - ['PMID: 22510445', '24700879', '14730302.']
TBC1D1  - ['PMID: 26572137']
TRAP1   - ['PMID: 24152966 - recessive mutations reported in 2 families with CAKUT, and 3 families with VACTERL.']
SALL4   - ['PMID: 20301547']
```
Most likely there is some issue with the regular expression cleaning these strings. However I found it strange that GRIP1 has only 2 pmids, indicating more than one were resolved to `2` then cased into a set.

Reply:
Thank you for noticing! It was indeed due to a regex issue. Now corrected, & also slightly improved the post-filtering of PMIDs: https://github.com/opentargets/evidence_datasource_parsers/pull/88/commits/ac83edb4f832f10552faaff10e96eb036e53f1a0